graph/formats/rdf/rdf.go

```diff
@@ -0,0 +1,384 @@
+// this by wrapping Statement in a type performing that operation.
+func (s *Statement) ReversedLine() graph.Line { return s }
+
+// ParseNQuad parses the statement and returns the corresponding Statement.
+// All Term UID fields are zero on return.
+func ParseNQuad(statement string) (*Statement, error) {
+	s, err := parse([]rune(statement))
+	if err != nil {
+		return nil, err
+	}
+	return &s, err
+}
+
+// Decoder is an RDF stream decoder. Statements returned by calls to the
+// Unmarshal method have their Terms' UID fields set so that unique terms
+// will have unique IDs and so can be used directly in a graph.Multi, or
+// in a graph.Graph if all predicate terms are identical. IDs created by
+// the decoder all exist within a single namespace and so Terms can be
+// uniquely identified by their UID. Term UIDs are based from 1 to allow
+// RDF-aware client graphs to assign ID if no ID has been assigned.
+type Decoder struct {
+	scanner *bufio.Scanner
+
+	strings store
+	ids     map[string]int64
+}
+
+// NewDecoder returns a new Decoder that takes input from r.
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{
+		scanner: bufio.NewScanner(r),
+		strings: make(store),
+		ids:     make(map[string]int64),
+	}
+}
+
+// Reset resets the decoder to use the provided io.Reader, retaining
+// the existing Term ID mapping.
+func (dec *Decoder) Reset(r io.Reader) {
+	dec.scanner = bufio.NewScanner(r)
+	dec.strings = make(store)
+	if dec.ids == nil {
+		dec.ids = make(map[string]int64)
+	}
+}
+
+// Unmarshal returns the next statement from the input stream.
+func (dec *Decoder) Unmarshal() (*Statement, error) {
+	for dec.scanner.Scan() {
+		data := bytes.TrimSpace(dec.scanner.Bytes())
+		if len(data) != 0 && data[0] != '#' {
+			s, err := ParseNQuad(string(data))
+			if err != nil {
+				return nil, fmt.Errorf("rdf: failed to parse %q: %w", data, err)
+			}
+			if s == nil {
+				continue
+			}
+			s.Subject.Value = dec.strings.intern(s.Subject.Value)
+			s.Predicate.Value = dec.strings.intern(s.Predicate.Value)
+			s.Object.Value = dec.strings.intern(s.Object.Value)
+			s.Subject.UID = dec.idFor(s.Subject.Value)
+			s.Object.UID = dec.idFor(s.Object.Value)
+			s.Predicate.UID = dec.idFor(s.Predicate.Value)
+			if s.Label.Value != "" {
+				s.Label.Value = dec.strings.intern(s.Label.Value)
+				s.Label.UID = dec.idFor(s.Label.Value)
+			}
+			return s, nil
+		}
+	}
+	dec.strings = nil
+	err := dec.scanner.Err()
+	if err != nil {
+		return nil, err
+	}
+	return nil, io.EOF
+}
+
+func (dec *Decoder) idFor(s string) int64 {
+	id, ok := dec.ids[s]
+	if ok {
+		return id
+	}
+	id = int64(len(dec.ids)) + 1
+	dec.ids[s] = id
+	return id
+}
+
+// Terms returns the mapping between terms and graph node IDs constructed
+// during decoding the RDF statement stream.
+func (dec *Decoder) Terms() map[string]int64 {
+	return dec.ids
+}
+
+// store is a string internment implementation.
+type store map[string]string
```

Comment:
This is not the best approach to string interning. I am waiting to see what happens in the json package and the approach that ends up being used there and will probably base the approach on that.